zookeeper/src/main/java/com/linecorp/armeria/client/endpoint/zookeeper/common/Connector.java

```diff
@@ -0,0 +1,261 @@
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.zookeeper.KeeperException.Code.get;
+
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CountDownLatch;
+
+import org.apache.zookeeper.AsyncCallback.StatCallback;
+import org.apache.zookeeper.KeeperException.Code;
+import org.apache.zookeeper.WatchedEvent;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.Watcher.Event.KeeperState;
+import org.apache.zookeeper.ZooKeeper;
+import org.apache.zookeeper.data.Stat;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import com.linecorp.armeria.client.endpoint.EndpointGroupException;
+
+/**
+ * A ZooKeeper connector, maintains a ZooKeeper connection.
+ */
+public abstract class Connector {
+
+    private static final Logger logger = LoggerFactory.getLogger(Connector.class);
+
+    private static final int MAX_RETRY_DELAY_MILLIS = 60 * 1000; // One minute
+    private final String zkConnectionStr;
+
+
+
+    private final String zNodePath;
+    private final int sessionTimeout;
+    private int retryDelayMills = 1000; // Start with one second
+    private ZooKeeper zooKeeper;
+    private BlockingQueue<KeeperState> stateQueue;
+    private CountDownLatch latch;
+    private boolean activeClose;
+
+    /**
+     * Creates a connector
+     *
+     * @param zkConnectionStr a connection string containing a comma separated list of {@code host:port} pairs,
+     *                        each corresponding to a ZooKeeperProxy server
+     * @param zNodePath       a zNode path e.g. {@code "/groups/productionGroups"}
+     * @param sessionTimeout  Zookeeper session timeout in milliseconds
+     */
+    protected Connector(String zkConnectionStr, String zNodePath, int sessionTimeout
+    ) {
+
+        this.zkConnectionStr = requireNonNull(zkConnectionStr, "zkConnectionStr");
+        this.zNodePath = requireNonNull(zNodePath, "zNodePath");
+        this.sessionTimeout = sessionTimeout;
+
+    }
+
+    /**
+     * Do connect.
+     */
+    protected void connect() {
+
+        try {
+            activeClose = false;
+            latch = new CountDownLatch(1);
+            zooKeeper = new ZooKeeper(zkConnectionStr, sessionTimeout, new ZkWatcher());
+            latch.await();
+            postConnected(zooKeeper);
+            if (stateQueue != null) {
+                //put a fake stat to ensure method postConnected finished completely
+                //(so that it won't throw exception under ZooKeeper connection recovery test)
+                stateQueue.put(KeeperState.Disconnected);
+            }
+
+        } catch (Exception e) {
+            throw new ZooKeeperException(
+                    "failed to connect to ZooKeeper:  " + zkConnectionStr + " (" + e + ')', e);
+        }
+
+    }
+
+    protected void resetRetryDelay() {
+        retryDelayMills = 1000;
```

Comment:
`1000` could be extracted into a constant to remove the duplication: https://github.com/line/armeria/pull/357/files#diff-11d50d092013955c3b020163c3de7578R53